### ISSUE TITLE
contributing DiffuRank and Cap3D to allthingsearch

### ISSUE THREAD

**Issue by himerosai** [tiangeluo/DiffuRank#3]
Hi there,
I am have been trying to reach out to your team via emails.
I am the author of this project: https://github.com/himerosai/allthingsearch
which is a 3d search engine and would love to add DiffuRank and Cap3D for the captioning search.

Twitter: https://x.com/himerosai

Would love to discuss further.

**Comment by tiangeluo** [comment]
Hi @himerosai ,

Thank you for reaching out! Sorry for any inconvenience -- I may missed your email.

Love to discuss further around the 3d search engine! I am unsure about what is adding DiffuRank and Cap3D for the captioning search. Could you please elaborate more here or via email?

**Comment by priamai** [comment]
Hi @tiangeluo, 
sure, so referring to your paper documentation:

![image](https://github.com/tiangeluo/DiffuRank/assets/57333254/eaa474c6-d360-468c-8180-2deba136885c)

the user click upload on the web interface, select a 3D file (obj or glb format), we render the file in the browser, we then take the 6 views image screenshots (I save them in the database), then we run the captioning Blip2 (I save the text in the database), then the GPT4 vision to summarize them (I save the text in the database).

That way then the user can simply type in the search bar the description of the object he/she wants, then we do some clever filtering/ranking based on traditional keyword matching (via Elasticsearch) and semantic embeddings (ADA embeddings) via vector search.

I think we can start with Cap3D first, then we do DiffuRank which to me seems a bit more complex.

**Comment by priamai** [comment]
Currently I am adding your pipeline:

https://github.com/crockwell/Cap3D/tree/main/captioning_pipeline

which I need to trigger for each file upload.

**Comment by priamai** [comment]
I am getting this error on my Ubuntu (headless) server:

```
3d@all-3d-index:~/Cap3D/captioning_pipeline$ ./blender-3.4.1-linux-x64/blender -b -P render_script.py -- --object_path_pkl './example_material/example_object_path.pkl' --parent_dir './example_material'
./blender-3.4.1-linux-x64/blender: error while loading shared libraries: libxkbcommon.so.0: cannot open shared object file: No such file or directory
```
I am not sure what that library is, any suggestions?

**Comment by priamai** [comment]
Updates, I managed to fix it with a simple:

`sudo apt-get install libxkbcommon-x11-0`

or on Fedora:

`yum install libxkbcommon-x11`

**Comment by priamai** [comment]
Question: what should we do to improve the processing latency?

I noticed it took 1 minute 58 seconds for the example:

```
Saved: './example_material/Cap3D_imgs/Cap3D_imgs_view7/d1e62f1eb6334a1196b53d61529f472b_7.png'
 Time: 00:01.58 (Saving: 00:00.03)
```
I need to add Celery to manage background tasks.

**Comment by tiangeluo** [comment]
> Hi @tiangeluo, sure, so referring to your paper documentation:
> 
> ![image](https://private-user-images.githubusercontent.com/57333254/343375675-eaa474c6-d360-468c-8180-2deba136885c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTk2OTczODEsIm5iZiI6MTcxOTY5NzA4MSwicGF0aCI6Ii81NzMzMzI1NC8zNDMzNzU2NzUtZWFhNDc0YzYtZDM2MC00NjhjLTgxODAtMmRlYmExMzY4ODVjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjI5VDIxMzgwMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWIxYTI1ZGU2YjJhNjRiNDMwMDYzZjRmZmQwMjYzNWNhMDkyOTBiZDJhMTEyZjFkYTZkYTJlMmM3MTM5ODcxNmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.itPKcMQeWp5BL_BS4aIbQxnyfmhyys8_RNbt2TR6nv4)
> 
> the user click upload on the web interface, select a 3D file (obj or glb format), we render the file in the browser, we then take the 6 views image screenshots (I save them in the database), then we run the captioning Blip2 (I save the text in the database), then the GPT4 vision to summarize them (I save the text in the database).
> 
> That way then the user can simply type in the search bar the description of the object he/she wants, then we do some clever filtering/ranking based on traditional keyword matching (via Elasticsearch) and semantic embeddings (ADA embeddings) via vector search.
> 
> I think we can start with Cap3D first, then we do DiffuRank which to me seems a bit more complex.

This sounds interesting! 

Regarding "what should we do to improve the processing latency?", have you used this script: https://github.com/tiangeluo/DiffuRank/blob/main/render_script_type1.py? Knowing which script you used will help me potentially accelerate the entire process.

Another thought: It might be beneficial for the website to initially support browser Objaverse objects. This would allow for the integration of rendered images, extracted point clouds, captions, and other information already available in the Cap3D-huggingface repository.

**Comment by himerosai** [comment]
Yes as I mentioned earlier in my post, I just ran the default code:
```

3d@all-3d-index:~/Cap3D/captioning_pipeline$ ./blender-3.4.1-linux-x64/blender -b -P render_script.py -- --object_path_pkl './example_material/example_object_path.pkl' --parent_dir './example_material'
```

I am not familiar with blender so I would like to understand what is the relation between the blender modified binary and the render_script.py.

Does Blender basically execute that script internally?

Also I am a bit confused about the example_object_path.pkl file: is it just to store which uid files needs to be processed?

**Comment by tiangeluo** [comment]
Hi @himerosai , 

Very sorry for the late response due to traveling. 

> Does Blender basically execute that script internally?

Yeah, the python of Blender will execute the script in the background. Just like `python xx.py` but using Blender python.

> is it just to store which example_object_path.pkl files needs to be processed?

It stores the path to the object file to be processed. 

> I just ran the default code

Could you please try this code, https://github.com/tiangeluo/DiffuRank/blob/main/render_script_type1.py, for speed?
